### PR TITLE
Add Transformerunittest

### PR DESCRIPTION
**Pull request by gongweibao**
Add dist transformer unit test.

Fix https://github.com/PaddlePaddle/Paddle/issues/13007

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/dist_transformer.py [hunk 1]
```diff
@@ -205,63 +130,89 @@ def _wait_ps_ready(self, pid):
...
-            WMT16_RECORDIO_FILE) as writer:
-        for batch in reader():
-            for tensor in prepare_batch_input(
-                    batch, ModelHyperParams.src_pad_idx,
-                    ModelHyperParams.trg_pad_idx, ModelHyperParams.n_head):
-                t = fluid.LoDTensor()
-                t.set(tensor, fluid.CPUPlace())
-                writer.append_tensor(t)
-            writer.complete_append_tensor()
+         is_dist=True,
```
Comment (typhoonzero) [inline python/paddle/fluid/tests/unittests/dist_transformer.py]: Should use `runtime_main` in test_dist_base.py

File: python/paddle/fluid/tests/unittests/optim.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+# Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (typhoonzero) [inline python/paddle/fluid/tests/unittests/optim.py]: Put all these contents to `dist_transformer.py` to avoid too many files under `unitests` folder.

Comment (gongweibao) [reply python/paddle/fluid/tests/unittests/optim.py]: Done.

File: python/paddle/fluid/tests/unittests/dist_transformer.py (hunk 1 above)
Comment (gongweibao) [reply python/paddle/fluid/tests/unittests/dist_transformer.py]: Done.

**Comment by panyx0718** [comment]
What's the reason for this change? It seems there was a transfomer test? Can you explain in your PR description?

**Comment by gongweibao** [comment]
@panyx0718 I add an issue about this.

File: .gitignore [hunk 3]
```diff
@@ -9,6 +9,7 @@ python/paddle/v2/fluid/tests/book/label_semantic_roles.inference.model/
 build/
 build_doc/
 *.user
+paddle/fluid/operators/distributed/send_recv.proto
```
Comment (typhoonzero) [inline .gitignore]: Should generate the file in the source directory, put it in the build directory. Can update this in another PR.

File: python/paddle/fluid/tests/unittests/test_dist_transformer.py [hunk 4]
```diff
@@ -15,17 +15,53 @@
...
+class TestDistTransformer2x2Sync(TestDistBase):
     def _setup_config(self):
         self._sync_mode = True
 
     def test_transformer(self):
-        # TODO(paddle-dev): check if the delta is OK.
-        # Usually start around ~8000 and converge to ~5000
-        self.check_with_place("dist_transformer.py", delta=400)
+        download_files()
+        self.check_with_place("dist_transformer.py", delta=0.00001)
```
Comment (typhoonzero) [inline python/paddle/fluid/tests/unittests/test_dist_transformer.py]: Can delta be fixed to 1e-7? why

File: python/paddle/fluid/tests/unittests/test_dist_transformer.py [hunk 5]
```diff
@@ -15,17 +15,53 @@
...
+
+
+class TestDistTransformer2x2Async(TestDistBase):
+    def _setup_config(self):
+        self._sync_mode = False
+
+    def test_transformer(self):
+        download_files()
+        self.check_with_place("dist_transformer.py", delta=0.1)
```
Comment (typhoonzero) [inline python/paddle/fluid/tests/unittests/test_dist_transformer.py]: Can increase async mode delta value, or this test will fail sometime.

**Review by typhoonzero — APPROVED** [review]

**Pull request merged**